google/cloud/internal/future_impl.h

```diff
@@ -0,0 +1,198 @@
+   *
+   * @param deadline the maximum time to wait.
+   *
+   * @tparam Clock a placeholder to match the Clock tparam for @p tp's
+   *     type, the semantics of this template parameter are documented in
+   *     `std::chrono::time_point<>` (in brief, the underlying clock type
+   *     associated with the time point), for our purposes it is simply a
+   *     formal parameter.
+   *
+   * @return `std::future_status::ready` if the shared state is satisfied.
+   *     `std::future_status::deferred` if the shared state is not satisfied and
+   *     there is a continuation ready to execute when it is satisfied.
+   *     `std::future_status::timeout` otherwise.
+   */
+  template <typename Clock>
+  std::future_status wait_until(std::chrono::time_point<Clock> deadline) {
+    std::unique_lock<std::mutex> lk(mu_);
+    if (not lk.owns_lock()) {
+      return std::future_status::timeout;
+    }
+    bool result =
+        cv_.wait_until(lk, deadline, [this] { return is_ready_unlocked(); });
+    if (result) {
+      return std::future_status::ready;
+    }
+    return std::future_status::timeout;
+  }
+
+  /// Set the shared state to hold an exception and notify immediately.
+  void set_exception(std::exception_ptr ex) {
+    std::unique_lock<std::mutex> lk(mu_);
+    set_exception(std::move(ex), lk);
+    notify_now(lk);
+  }
+
+  /**
+   * Abandon the shared state.
+   *
+   * The destructor of `promise<T>` abandons the state. If it is satisfied this
+   * has no effect, but otherwise the state is satisfied with an
+   * `std::future_error` exception. The error code is
+   * `std::future_errc::broken_promise`.
+   */
+  void abandon() {
+    std::unique_lock<std::mutex> lk(mu_);
+    if (is_ready_unlocked()) {
+      return;
+    }
+    set_exception(std::make_exception_ptr(
+                      std::future_error(std::future_errc::broken_promise)),
+                  lk);
+    notify_now(lk);
+  }
+
+ protected:
+  bool is_ready_unlocked() const { return current_state_ != state::not_ready; }
+
+  /// Satisfy the shared state using an exception.
+  void set_exception(std::exception_ptr ex, std::unique_lock<std::mutex>& lk) {
+    if (is_ready_unlocked()) {
+      throw std::future_error(std::future_errc::promise_already_satisfied);
+    }
+    exception_ = std::move(ex);
+    current_state_ = state::has_exception;
+  }
+
+  /// Immediately notify all waiting threads.
+  void notify_now(std::unique_lock<std::mutex>& lk) {
+    // In the future we may implement notify_at_thread_exit() to notify waiters
+    // only when the current thread exits.
+    cv_.notify_all();
+    // Release the lock after the notification because otherwise the threads
+    // may lose the state change.
+    lk.unlock();
+  }
+
+  mutable std::mutex mu_;
+  std::condition_variable cv_;
+  enum class state {
+    not_ready,
+    has_exception,
+    has_value,
+  };
+  state current_state_;
+  std::exception_ptr exception_;
```

Comment:
If we used C++17 I'd have suggested a `std::variant` for `state_` and `exception_`, but it's probably not worth the time to hand-craft it.